mera/utils.py

```diff
@@ -0,0 +1,133 @@
+        A list of warning strings for records with an insufficient
+        number of records per site or per event.
+
+    """
+    count = residual_df_per_im.groupby(group_col).count().iloc[:, 0]
+    mask = count < min_num_group_warn
+
+    warn_str_lines = [
+        f"{count.loc[mask].index[x]} has only {count.loc[mask].iloc[x]} records "
+        f"(recommended minimum is {min_num_group_warn})"
+        for x in range(len(count.loc[mask]))
+    ]
+
+    return warn_str_lines
+
+
+def mask_too_few_records(
+    residual_dataframe: pd.DataFrame,
+    event_cname: str = "event_id",
+    site_cname: str = "stat_id",
+    mask: Optional[pd.DataFrame] = None,
+    min_num_records_per_event: int = 3,
+    min_num_records_per_site: int = 3,
+) -> pd.DataFrame:
+    """
+    Creates a boolean mask (or modifies one passed in) that removes records without sufficient records per station/event.
+    Note: This is done independently per IM.
+
+    Parameters
+    ----------
+    residual_dataframe: DataFrame
+        Residual DataFrame, has to contain all
+        specified IMs (as columns) along with
+         columns for event and site id.
+    mask: DataFrame, optional
+        Mask DataFrame including the same IM columns as residual_df but not including the site and event id columns
+        (the shape of the mask DataFrame should be the same as residual_df[ims]).
+    event_cname: string, default = "event_id"
+        Name of the column that contains the event ids.
+    site_cname: string, default = "stat_id"
+        Name of the column that contains the site ids.
+    min_num_records_per_event : int, default = 3
+        The minimum number of records per event required to keep the records.
+    min_num_records_per_site : int, default = 3
+        The minimum number of records per site required to keep the records.
+
+    Returns
+    -------
+    mask: pd.DataFrame
+        Mask DataFrame with the same shape as the given residual DataFrame.
+    """
+
+    # making a copy to ensure that the original DataFrame is not modified
+    residual_df = residual_dataframe.copy()
+
+    ims = residual_df.columns.drop([event_cname, site_cname])
+
+    if mask is None:
+        # This mask must always have the event_cname and site_cname columns as all True
+        # so residual_df[mask] includes those columns for groupby (below)
+        mask = residual_df[ims].notnull()
+
+    # To mask records based on the number of records per station or event, an iterative method is needed.
+    # For example, if all records of a particular event are masked, there may then be an insufficient
+    # number of records from an affected site, so they need to be masked out too.
+    # Therefore, this code iterates until the mask is not changed by an additional iteration.
+
+    iteration_counter = 0
+
+    # Add the event and site id columns to the mask so that the masked DataFrame will also have those columns for groupby
+    mask[event_cname] = True
```

Comment:
I don't think these are needed. 
Should just need to change
`residual_df[mask].groupby(event_cname).transform("count")[ims]`
to 
`residual_df[ims][mask].groupby(event_cname).transform("count")`

Reply:
It might be best to discuss this in person to see if I properly understand what you're suggesting. I think in what you've suggested, mask will still need to include event_cname and site_cname so that groupby can see those columns after masking

Reply:
Yeah you are right, what I suggested doesn't work.